### ISSUE TITLE
Improve ES write throughput

### ISSUE THREAD

**Issue by polygloton** [threatgrid/ctia#442]
See description in my comment below:

**Comment by polygloton** [comment]
Originally reported by @saintx, loading lots of data into CTIA (backed by ES) is slow, especially compared to loading directly into ES.  Loading into CTIA results in about 70 insertions per second when using the bulk API.  Loading data directly into ElasticSearch (bypassing CTIA) is significantly faster.

I profiled CTIA during a data import* to assess what was happening.  I found that 96% of CPU time was spent in calls to clj-http.client/post.  The time was split up between two such *post* calls.  52% of the time was taken up while inserting new entities into ElasticSearch and the other 44% was taken up while inserting events into ElasticSearch.  Actual CPU utilization on my laptop was low (rarely peaking at 11%) and memory utilization and GC was low.  That means that most of the time the CTIA threads were asleep waiting for ES to finish.

Currently in CTIA, the *post* calls to ES are made sequentially within a single web-handler thread.  When we are inserting a single entity into CTIA, there are two *post* calls for ES.  When the CTIA bulk API is used, each entity in the bulk payload is sent to ES sequentially.  This is the cause of the slowness that @saintx  is seeing.

@saintx, @gbuisson, and I discussed two approaches for improving the situation.

Alternative approaches:

- Do bulk insertions into ES using the ES bulk API
  - Assumes that fewer, larger *post* calls will perform better than lots of small ones
- For the CTIA bulk API, make parallel insertions into ES
  - Assumes that ES can handle lots of parallel requests (supported by @saintx's testing)

My hope is that we can do both (they are not mutually exclusive).

Footnote:
* Test was done with local CTIA and ES using threatgrid/hydrant

**Comment by craigbro** [comment]
Do the first, using BULK insert API in ES.

**Comment by craigbro** [comment]
Here is some example code illustrating how we do that in Kiries now, using the elastich bulk API.

https://github.com/threatgrid/kiries/blob/master/src/riemann/elastic.clj#L96

**Comment by polygloton** [comment]
Inserting entities into ES is now done using the ES bulk API.  However, inserting events into ES is still done using event hooks and therefore are inserted one at a time.  My next task is to do bulk event insertions into ES.

**Comment by polygloton** [comment]
I made [an attempt](https://github.com/polygloton/ctia/tree/issue-442-b-event-store) to abstract writing events into an event store.  However, I found that the abstraction of the stores didn't match that of the event stream.  Mainly, the problem is that events do not have IDs.  The closest match to an ID is the ID of the entity which the event is for, however, it can not be used since there will be multiple events for a given entity instance.  I decided to abandon that approach and go with the much simpler solution of moving the ES event writing code that currently exists into the flow code.  This approach is simpler because it doesn't require me to also get rid of the producer abstraction; it uses that code as-is.
